File: ghost-env/src/replay.ts
import type { CallRecord } from "./recording/recorder.js";

/** Replay saved responses in order (record/export roundtrip). */
export class ReplayFixture {
  private idx = 0;
  constructor(private readonly calls: CallRecord[]) {}

  static fromJSON(json: string): ReplayFixture {
    return new ReplayFixture(JSON.parse(json) as CallRecord[]);
  }

  nextResponse(): Response | null {
    const c = this.calls[this.idx];
    if (!c) return null;
    this.idx += 1;
    const body =
      typeof c.responseBody === "string" ? c.responseBody : JSON.stringify(c.responseBody ?? "");
    return new Response(body, { status: c.responseStatus, headers: { "Content-Type": "application/json" } });
  }

  reset(): void {
    this.idx = 0;
  }
}

File: ghost-env/python/src/ghost_env/replay.py
from __future__ import annotations

import json
from typing import Any

from ghost_env.recorder import CallRecord


class ReplayFixture:
    def __init__(self, calls: list[dict[str, Any]]) -> None:
        self._calls = calls
        self._idx = 0

    @classmethod
    def from_json(cls, raw: str) -> ReplayFixture:
        data = json.loads(raw)
        return cls(data)

    def next_response(self) -> tuple[int, str] | None:
        if self._idx >= len(self._calls):
            return None
        c = self._calls[self._idx]
        self._idx += 1
        body = c.get("response_body")
        text = body if isinstance(body, str) else json.dumps(body)
        return int(c.get("response_status", 200)), text

    def reset(self) -> None:
        self._idx = 0
